### ISSUE TITLE
BackgroundFetch not running

### ISSUE THREAD

**Issue by leighquince** [expo/expo#3582]
### Environment
```
Expo CLI 2.11.1 environment info:
    System:
      OS: macOS 10.14.3
      Shell: 5.3 - /bin/zsh
    Binaries:
      Node: 10.6.0 - /usr/local/bin/node
      Yarn: 1.12.3 - /usr/local/bin/yarn
      npm: 6.8.0 - /usr/local/bin/npm
      Watchman: 4.9.0 - /usr/local/bin/watchman
    IDEs:
      Xcode: 10.1/10B61 - /usr/bin/xcodebuild
    npmGlobalPackages:
      expo-cli: 2.11.1
```
 app's target - iOS, Standalone

### Steps to Reproduce
1. Allow background fetch in app.json
```
"ios": {
    "infoPlist": {
        "UIBackgroundModes": [
		"location",
		"fetch"
	]
```
2. define a task in App.js
 ```
defineHealthTask = () => {
		TaskManager.defineTask(taskName, async () => {
			try {
				await UserRemoteLogService.logWithTaskType("STEPS_TASK_RUNNING");
				return BackgroundFetch.Result.NewData;
			} catch (error) {
			        return BackgroundFetch.Result.Failed;
			}
		});
	};
```
3. register task at some point of the apps running

```
await UserRemoteLogService.logWithTaskType("STEPS_TASK_REGISTRATION_REQUESTED");
	const pedometerAvailable = await Pedometer.isAvailableAsync();
	const backGroundFetchStatus = await BackgroundFetch.getStatusAsync();
	if (pedometerAvailable && BackgroundFetch.Status.Available === backGroundFetchStatus) {
		await BackgroundFetch.registerTaskAsync(taskName, {
			userBusinessId: user.businessId
		});
		UserRemoteLogService.logWithTaskType("STEPS_TASK_REGISTERED");
	}
```
4. set minimum interval for task in seconds
`BackgroundFetch.setMinimumIntervalAsync(60);`

### Expected Behavior

App should periodically call my server to log that the background task is running

### Actual Behavior

Never see the background task run, although i do see the logs that show the background task was registered. The logging service i use has been used to show when location background task has run so this logging service should log if the task runs

My main question is has anyone seen background fetch run? if so have i just configured it wrong?

**Comment by sjchmiela** [comment]
Hey, @leighquince! As noted in [the documentation](https://docs.expo.io/versions/v32.0.0/sdk/background-fetch/#backgroundfetchregistertaskasynctaskname), `TaskManager.defineTask` has to be called in global scope of the app, not in a component's function (like `defineHealthTask`). Try moving call to `defineTask` to the global level of the app (like in the example in the documentation) and see if it helps.

**Comment by leighquince** [comment]
In my `App.js` I call `defineHealthTask`. Calling it directly in `App.js` or via a function that would be evaluated at the same time should have no difference? I'll give it a go though, thanks

**Comment by tcdavis** [comment]
Would help if you could give us a snack or toy repo demonstrating the issue if changing the call context doesn't resolve it

**Comment by leighquince** [comment]
A quick snack https://snack.expo.io/@domosafety/background-fetch. 

Task is defined at top level, registered when App mounts, no logs are seen to show that the background task runs

**Comment by Oakyden** [comment]
Struggling with this too, i have defined a task in Global scope (GET_STEPS), but console logging the promise returns "undefined".

`BackgroundFetch.registerTaskAsync(GET_STEPS)
.then(
  (data) => {console.log("task data is ", data)}
)`

**Comment by superyarik** [comment]
@Oakyden there is no data income in BackgroundFetch. you can call fetch, process it and return BackgroundFetch.Result.### result.

**Comment by superyarik** [comment]
i can confirm that BackgroundFetch work for me, but very strange. looks like i can not call anything like in main app(asyncstorage, get coordinates or something like this) and frequency of calling very various, it's making debug hard. So i decide to wait android implementation.

**Comment by Oakyden** [comment]
Can you share an example expo snack or similar @superyarik ?  especially if you use a simple API example like https://jsonplaceholder.typicode.com/

**Comment by mskg** [comment]
It took me a long while to understand what was my problem. The expo documentation is a bit misleading. According to [Apple](https://developer.apple.com/documentation/uikit/uiapplication/1623100-setminimumbackgroundfetchinterva) you **MUST** call `BackgroundFetch.setMinimumIntervalAsync` in order for the job to be scheduled. 

So, you have to 
1. Always call `TaskManager.defineTask`
2. Check if task is registered with `BackgroundFetch`, if not, register
3. Always call `BackgroundFetch.setMinimumIntervalAsync`

Be aware that an emulator does not run your task, that happens only on real devices.

The following worked for me:
```typescript
export async function registerFetchTask() {
    TaskManager.defineTask(FETCH_TASKNAME, runBackgroundSaga);

    const status = await BackgroundFetch.getStatusAsync();
    switch (status) {
        case BackgroundFetch.Status.Restricted:
        case BackgroundFetch.Status.Denied:
            logger.log("Background execution is disabled");
            return;

        default: {
            logger.debug("Background execution allowed");

            let tasks = await TaskManager.getRegisteredTasksAsync();
            if (tasks.find(f => f.taskName === FETCH_TASKNAME) == null) {
                logger.log("Registering task");
                await BackgroundFetch.registerTaskAsync(FETCH_TASKNAME);

                tasks = await TaskManager.getRegisteredTasksAsync();
                logger.debug("Registered tasks", tasks);
            } else {
                logger.log(`Task ${FETCH_TASKNAME} already registered, skipping`);
            }

            logger.log("Setting interval to", INTERVAL);
            await BackgroundFetch.setMinimumIntervalAsync(INTERVAL);
        }
    }
}
```

**Comment by choco9620** [comment]
Hello, I have the same problem, the global task was defined but it is not executed in the interval, is it possible that someone can share functional code ?, since the expo documentation does not define anything.

**Comment by tomas1000r** [comment]
Hello, it looks like we need to use `Location.startLocationUpdatesAsync('taskName')` in order to get background tasks working. It also requires that we ask for the location permission and it should be set to "Allow always". It looks like that BackgroundFetch is tightly coupled with location. Don't know why.

**Comment by BrodaNoel** [comment]
@mskg you should call `BackgroundFetch.setMinimumIntervalAsync(INTERVAL);` only if you are registering the app, right?

**Comment by mskg** [comment]
@BrodaNoel no, you MUST call it every time when the App starts. See https://developer.apple.com/documentation/uikit/uiapplication/1623100-setminimumbackgroundfetchinterva

**Comment by SowinskiMateusz** [comment]
> Hello, it looks like we need to use `Location.startLocationUpdatesAsync('taskName')` in order to get background tasks working. It also requires that we ask for the location permission and it should be set to "Allow always". It looks like that BackgroundFetch is tightly coupled with location. Don't know why.

Is it really true? This is the only command that fires the task? Does not it make sense? Is there any other command not related to Location?

**Comment by mskg** [comment]
@SowinskiMateusz no, see my example above.

**Comment by Bug-Reaper** [comment]
@mskg

Could you provide a more complete example, or possibly add comments to your original

**Comment by farzd** [comment]
Doesnt appear to work for me either, my setup for geofencing works so following same convention, define task in global scope etc

```
  TaskManager.defineTask(client.tasks.background, () => {
    try {
      console.log('doesnt log')
      return BackgroundFetch.Result.NewData
    } catch (error) {
      return BackgroundFetch.Result.Failed
    }
  })
```

```
    let isRegistered = await TaskManager.isTaskRegisteredAsync(
      client.tasks.background
    )
    if (isRegistered) {
      console.log('unregister task ...')
      await BackgroundFetch.unregisterTaskAsync(client.tasks.background)
    }

    const status = await BackgroundFetch.getStatusAsync()
    console.log(BackgroundFetch.Status[status]) //Available 
    await BackgroundFetch.setMinimumIntervalAsync(100)
    await BackgroundFetch.registerTaskAsync(client.tasks.background)
```

**Comment by farzd** [comment]
The example provided here: https://github.com/expo/expo/blob/master/apps/native-component-list/screens/BackgroundFetchScreen.js

I've tailored it to work without the alpha release and without react navigation,
it registers the task and task status is available but no logs for fetch [tested on iPhone7, ios12, via expo client]

code: https://pastebin.com/gNU5QnJv

**Comment by tsapeta** [comment]
> So, you have to
> 
> 1. Always call `TaskManager.defineTask`.
> 2. Check if task is registered with `BackgroundFetch`. If not, register.
> 3. Always call `BackgroundFetch.setMinimumIntervalAsync`.

re: 2. - there is no need to check if background fetch task is already registered - if you try to register it again, then it just overwrites the previously registered task, which is fine 👌 Except that, these steps look good.

> adding this at the end appears to make it work, as suggested by @tomas1000r why? @tsapeta
> 
> ```
>     Location.startLocationUpdatesAsync(client.tasks.background)
> ```

I'll try to check it out soon, but it seems weird to me as there is no connection between location and background fetch. They are completely independent features. My assumption is that, if you turn on background location then your app is actually always active, so the background fetch can occur in the time you set by `setMinimumIntervalAsync`. However, if the app is not active, then it strongly depends on the operating system and how you use the application. Apple has its own algorithm which defines how often the background fetch should trigger, based on your own usage of the app. If you use it a lot, then it will fetch as often as possible, but if you use it like at 4pm every day, the background fetch should trigger just before, so your data is updated when you launch it. So imho you shouldn't worry about the task not being done on time - Apple knows better when to call it 😅

**Comment by farzd** [comment]
@tsapeta can you please explain why this doesnt work: https://pastebin.com/gNU5QnJv
its taken from your master branch and altered slightly so updates are shown when app is brought to foreground, the fetch call is never made.

**Comment by tsapeta** [comment]
@farzd 
In this pastebin, you're passing a second argument with task options to `BackgroundFetch.registerTaskAsync` that is not yet available in Expo Client. It will be once we release SDK33. In SDK32 the minimum interval must be set by `BackgroundFetch.setMinimumIntervalAsync` - and it must be called because iOS defaults to a really big number.
Also, keep in mind that BackgroundFetch is not supported on Android in SDK32.

**Comment by farzd** [comment]
thanks for your time @tsapeta, i realised the arguments were for SDK33 but assumed it would no op anyway.
I've tried the code with explicitly setting ```BackgroundFetch.setMinimumIntervalAsync``` however it has made no difference, this is the issue a lot of people on this thread are facing.

Background app refresh is on in settings etc, not sure what else to check

**Comment by mskg** [comment]
At least according to my experience here: On iOS it is actually really like @tsapeta described. The time you set is just an indication on how often data might change. If it doesn’t (depending on the result), you‘re scheduled less frequent. Also, if you don’t use the App, the fetch will most likely not run.

In my case, the fetch currently runs directly after/when I open the App in the morning. I seem to do that regularly :)

**Comment by baptistemanson** [comment]
I directly tested the snack on a ios device and on a simulator using the expo app, and I cannot get 
```js
console.log('background fetch running') ;
``` 
to display, with or without remote debugging on, even after waiting 10min for it to occur. How can I debug further?

**Comment by g-bastianelli** [comment]
I m unable to make it works on ios with SDK33. Can have a working snack or example?

**Comment by mslavik2** [comment]
I am having trouble with SDK33 on ios as well. A working snack would be very helpful in debugging the issue and solidify the steps required to make the background fetch work.

**Comment by Kency1013** [comment]
BackgroundFetch not working with me. I use expo client test it. Is it only working in standalone app?

**Comment by amaurymartiny** [comment]
I'm using it as a standalone app on Android.

What I noticed is that when the app is not in the foreground but not terminated, i.e. in the [Recents Screen](https://developer.android.com/guide/components/activities/recents), the BackgroundFetch works well.

When the app is terminated (not in the Recents Screen), then BackgroundFetch never happens.

Is this an expected behavior?

Note : I manually set `stopOnTerminate` to false in the options.

**Comment by Hawods** [comment]
Mark. Waiting for the next version to close this issue.

**Comment by fhelwanger** [comment]
I'm also having the same problem as @amaurymartiny, where the background task is only executed when the app is in the Recents Screen. When the app is terminated the background task doesn't execute.

If it's not expected behavior I'd be happy to provide a simple snack to test.

**Comment by oguzhankarabulut** [comment]
@fhelwanger me too :(

**Comment by kpheasey** [comment]
Using Expo SDK 34 with a modified version solution from @mskg, I was able to get my tasks running in a stand alone app.  The basic idea is to call `BackgroundFetch.setMinimumIntervalAsync` AFTER the task has been registered.  While the original solution does have `await`, it was not clear to me that it was an essential part.

`TaskManager.defineTask` should still be called in the global scope.

```js
BackgroundFetch.registerTaskAsync(
    TASK_NAME,
    {
      minimumInterval: 60,
      stopOnTerminate: false,
      startOnBoot: true,
    },
).then(() => BackgroundFetch.setMinimumIntervalAsync(60));
```

BackgroundFetch does not work on iOS Simulator, you can only trigger a BackgroundFetch using Xcode.  So, unless you use ExpoKit, the only way to test iOS BackgroundFetch is to push a build to TestFlight!

**Comment by erin-truong** [comment]
I use Expo SDK 33.  I use expo client test with BackgroundFetch. The TaskManager is registered but never run.  Can have a working snack or example?

**Comment by niv-breez** [comment]
@superyarik did AsyncStorage work for you? what did you do in order to make it work?
Thanks

**Comment by NickR49** [comment]
@erin-truong Here's a working example for both iOS and Android (though Android is only running background fetch once a minute) - https://github.com/NickR49/background-fetch

**Comment by niv-breez** [comment]
I have several questions regrading this code:
1. Does it working for you (you are getting locations) both on Android and
IOS in killed mode ?
2. Are you able to use AsyncStorage in the background task when the app is
in killed mode ? this is for sending the server the user id with the
location

**Comment by NickR49** [comment]
Hi @niv-breez , no I haven't tested those two things. However you could take that code and add those things to test it out.

**Comment by NickR49** [comment]
Hi all,

What I'm finding is that I must have the `await Location.startLocationUpdatesAsync(BACKGROUND_FETCH_TASK);` code to make background fetch work for iOS. However when I use that then unregistering the task fails with `Invalid task consumer. Cannot unregister task with name 'background-fetch' because it is associated with different consumer class.`. I'm guessing this is because the task started as a `backgroundFetch` task type and got changed to a `location` task type.

Any progress working out the root cause @tsapeta ?

cheers
Nick

**Comment by holgersindbaek** [comment]
Is there any update on this?

I seem to have gotten it to work now, but the backgroundfetch only runs once or day (if at all) even though it's set to run every half an hour.

Asking people to always allow location tracking doesn't seem to be a viable option to me, since my app has nothing to do with a users location at all.

Is this really how it's supposed to work? If that's really the case, it's seems to me to be a pretty useless API?!

**Comment by jeromecornet** [comment]
This is likely due to the loader error I raised in https://github.com/expo/expo/issues/6290 and that I think ends up being the problem in https://github.com/expo/expo/issues/6017 as well.

But yes, as it stands I think the BackgroundFetch API is just broken

**Comment by rosnaib11** [comment]
Same problem, not working. I already test on standalone, it trigger only once. I'm using android. I don't know how this supposed to work because on android it requires foreground service to preserve and don't kill the process running on the background. Instead, I just use the `Location.startLocationUpdatesAsync` for alternative but that efficient. I need to run my task exactly every 1 minute.

**Comment by kpheasey** [comment]
@rosnaib11 
> I need to run my task exactly every 1 minute

As I understand it, it's not possible to run it every 1 minute.  The OS batches background jobs from multiple applications and runs them based on some internal scheduling.  The requested run time is simply a suggestion to the OS.

The [BackgroundFetch documentation](https://docs.expo.io/versions/latest/sdk/background-fetch/#backgroundfetchsetminimumintervalasyncminimuminterval) touches on this briefly.

> Sets the minimum number of seconds that must elapse before another background fetch can be initiated. This value is advisory only and does not indicate the exact amount of time expected between fetch operations.
This method doesn't take any effect on Android.

**Comment by TheHunterDog** [comment]
any fix up yet ? 
or some example code I can't really figure this out if its fixed

**Comment by jeromecornet** [comment]
No, I tested it on SDK36 and it's still broken

**Comment by tuzz** [comment]
This seems to work for me on Android in the simulator. I haven't tested on iOS.

I wrote a base class to handle defining and registering tasks, then extended it.

**background_task.js** (base class)

```javascript
import * as TaskManager from "expo-task-manager";
import * as BackgroundFetch from "expo-background-fetch";

const Result = BackgroundFetch.Result;

class BackgroundTask {
  static name() {
    throw new Error("Implement me");
  }

  // Make this method return true if new data was fetched.
  static async run() {
    throw new Error("Implement me");
  }

  static enable({ log=false } = {}) {
    const name = this.name();
    const options = { stopOnTerminate: false, startOnBoot: true };
    const interval = 15; // minutes;

    log && console.log(`Defining ${name} background task...`);
    TaskManager.defineTask(name, () => this.runSafely(name, log));

    return BackgroundFetch.registerTaskAsync(name, options).then(() => {
      BackgroundFetch.setMinimumIntervalAsync(interval);
      log && console.log(`Successfully registered ${name} background task`);
    });
  }

  static async runSafely (name, log) {
    log && console.log(`Running ${name} background task...`);

    try {
      const bool = await this.run();
      return bool ? Result.NewData : Result.NoData;

    } catch (error) {
      log && console.error(`Error running ${name} background task:`, error);
      return Result.Failed;
    }
  }
}

export default BackgroundTask;
```

**my_task.js** (sub class)

```javascript
import BackgroundTask from "./background_task";

class MyTask extends BackgroundTask {
  static name() {
    return "MyTask";
  }

  static async run() {
    // await fetch(...);
    return true;
  }
};

export default MyTask;
```

**App.js**

```javascript
import MyTask from "./my_task";

MyTask.enable();

// ...
```

I also wrote some Jest tests for the BackgroundTask base class.

**background_task_spec.js**

```javascript
import BackgroundTask from "../../app/tasks/background_task";

import * as TaskManager from "expo-task-manager";
import * as BackgroundFetch from "expo-background-fetch";

jest.mock("expo-task-manager");
jest.mock("expo-background-fetch");

class TestBackgroundTask extends BackgroundTask {
  static name() {
    return "task name";
  }

  static async run() {
    this.taskHasRun = true;
  }
}

describe("BackgroundTask", () => {
  beforeEach(() => {
    BackgroundFetch.registerTaskAsync.mockResolvedValue();
    TestBackgroundTask.taskHasRun = false;
  });

  it("defines a TaskManager task", () => {
    TestBackgroundTask.enable();

    expect(TaskManager.defineTask)
      .lastCalledWith("task name", expect.anything());
  });

  it("sets the TaskManager task to the run function of the class", () => {
    TestBackgroundTask.enable();

    const lastCall = TaskManager.defineTask.mock.calls[0];
    const definedTask = lastCall[1];

    expect(TestBackgroundTask.taskHasRun).toBe(false);

    definedTask();
    expect(TestBackgroundTask.taskHasRun).toBe(true);
  });

  it("registers the task with BackgroundFetch", () => {
    TestBackgroundTask.enable();

    expect(BackgroundFetch.registerTaskAsync)
      .lastCalledWith("task name", expect.anything());
  });

  it("runs the task even if the phone reboots or the app terminates", () => {
    TestBackgroundTask.enable();

    const expectedOptions = { startOnBoot: true, stopOnTerminate: false };

    expect(BackgroundFetch.registerTaskAsync)
      .lastCalledWith(expect.anything(), expectedOptions);
  });

  it("runs the task every 15 minutes", async () => {
    await TestBackgroundTask.enable();

    expect(BackgroundFetch.setMinimumIntervalAsync).lastCalledWith(15);
  });

  it("can optionally log to the console", () => {
    jest.spyOn(global.console, "log").mockImplementation();

    TestBackgroundTask.enable({ log: false });
    expect(console.log).not.toHaveBeenCalled();

    TestBackgroundTask.enable({ log: true });
    expect(console.log).toHaveBeenCalled();
  });
});
```

Hopefully this helps someone.

**Comment by jeromecornet** [comment]
@tuzz thanks for the code... it's not substantially different from the example I posted, so I'm a bit confused as to why it would behave differently.
When you say 'it works' can you share whether you mean that:
- the tests pass
- the task gets executed while the app is still in the recently used apps (i.e. backgrounded)
- the task gets executed after the app was killed (removed from the recently used apps)
- the task gets executed after the phone was rebooted but before the app was restarted

Please let us know (as well as expo SDK version android simulator version, etc...) to narrow down the difference.

**Comment by tuzz** [comment]
The tests pass and the task gets executed while the app is in the recently used apps.

I'm not sure about the other two cases, I'll check them tomorrow and do some testing on a real device.

```sh
$ expo --version
3.11.1

$ cat yarn.lock | grep ^expo@
expo@^35.0.0:
```

Android simulator:

<img width="580" alt="Screenshot 2020-01-07 at 17 55 50" src="https://user-images.githubusercontent.com/892251/71917022-11ef8300-3177-11ea-8929-1892de204dcd.png">

**Comment by jeromecornet** [comment]
Ah ok that makes sense... the problem is with the last 2 cases, i.e. background fetch while the app is not running. (it works when the app is backgrounded)
The bug seems to be related to the way the app gets loaded when it's triggered by the task service which causes the app to crash and the task to get removed (see #6290 for more details)

**Comment by tuzz** [comment]
You're right, it doesn't work for the last two cases - only when the app is suspended.

Thanks for clearing that up.

**Comment by jeromecornet** [comment]
@tsapeta any movement on this issue ? Let us know what we can do to help troubleshoot this issue.

**Comment by jeromecornet** [comment]
I've submitted a PR to the doc stating that the feature does not work on devices/simulators in #6767 , we can always switch it back if the issue is fixed

**Comment by brentvatne** [comment]
@tsapeta will investigate this after he wraps up his current project

**Comment by jeromecornet** [comment]
ok, thanks for the heads up @brentvatne

**Comment by BrodaNoel** [comment]
There is something that I still can't get:
Is it supposed to work well in the Expo App (iOS), when the app is backgrounded?
I have the app (in Expo App) opened in background, and it doesn't work.

Does it only work after creating & publishing a build?

**Comment by roei133** [comment]
@mskg can you help me with my code, please?
I use it with Expo and I did an example like yours but it doesn't work for me and I don't understand why.
I  just want that my app will run automatically in the background every 1 minute.
I call this task at the beginning of the app.js :

`import * as BackgroundFetch from 'expo-background-fetch';
import * as TaskManager from 'expo-task-manager';

const FETCH_TASKNAME = 'test_task'
const INTERVAL = 60

function test() {
    console.log('function is running')
}

export async function registerFetchTask() {
    TaskManager.defineTask(FETCH_TASKNAME, test);

    const status = await BackgroundFetch.getStatusAsync();
    switch (status) {
        case BackgroundFetch.Status.Restricted:
        case BackgroundFetch.Status.Denied:
            console.log("Background execution is disabled");
            return;

        default: {
            console.debug("Background execution allowed");

            let tasks = await TaskManager.getRegisteredTasksAsync();
            if (tasks.find(f => f.taskName === FETCH_TASKNAME) == null) {
                console.log("Registering task");
                await BackgroundFetch.registerTaskAsync(FETCH_TASKNAME);

                tasks = await TaskManager.getRegisteredTasksAsync();
                console.debug("Registered tasks", tasks);
            } else {
                console.log(`Task ${FETCH_TASKNAME} already registered, skipping`);
            }

            console.log("Setting interval to", INTERVAL);
            await BackgroundFetch.setMinimumIntervalAsync(INTERVAL);
        }
    }

}`

**Comment by roei133** [comment]
@mskg  please help me ..
see my example above .. it's the same as yours but it doesn't work for me.
what is my problem can be ?

**Comment by tsapeta** [comment]
> There is something that I still can't get:
> Is it supposed to work well in the Expo App (iOS), when the app is backgrounded?
> I have the app (in Expo App) opened in background, and it doesn't work.
> 
> Does it only work after creating & publishing a build?

@BrodaNoel: Yes, it's supposed to work in the Expo Client and you shouldn't need to build/publish your app, it should work in the development if `expo start` process is still running. However, please take into account that background fetch is not called very often on iOS so it's pretty hard to debug. The easiest way to trigger background fetch is to:
1. build and run Expo client on your own using source code from this repo
2. launch your app in the client
3. close (but don't kill) the client
4. press Debug -> Simulate Background Fetch in Xcode's menu (unfortunately this menu option is unavailable when the app is not launched by Xcode, so that's why you have to install the client manually)

@roei133: Looks like your code may be incorrect - when and where do you call `registerFetchTask ` function? `TaskManager.defineTask` that is inside your function must be called in the top-level scope, otherwise it won't be seen when your task is running in headless mode because it may not even reach this part of code. Also, looks like you expect the task to be called in every minute, but this API can't do what you want. Even if you use so small `minimumInterval` there is still no guarantee on how often it will actually run - it depends on Apple's internal algorithm.

**Comment by roei133** [comment]
@tsapeta : i call to registerFetchTask (); in the app.js in the top ..

this is my app.js code and i call to **registerFetchTask** on the beginning.
what is not good please let me know its so important .. 

```
import React from 'react';
import { StyleSheet, Text, View, Platform } from 'react-native';
import { AzureInstance, AzureLoginView } from 'react-native-azure-ad-2';
import PlacesNavigator from './navigation/PlacesNavigator';
import Modal from './components/Modal';
import { registerFetchTask } from './helpers/registerFetchTask'
import moment from 'moment'

registerFetchTask();


const credentials = {
  client_id: 'ea00ca9e-8c37-4520-8d80-2c2bb9239bf8',
  scope: 'User.Read',
};

export default class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      azureLoginObject: {},
      loginSuccess: false
    };
    this.azureInstance = new AzureInstance(credentials);
    this._onLoginSuccess = this._onLoginSuccess.bind(this);
  }


  _onLoginSuccess() {
    this.azureInstance.getUserInfo().then(result => {
      this.setState({
        loginSuccess: true,
        azureLoginObject: result,
      });
      console.log(result);
    }).catch(err => {
      console.log(err);
    })
  }

  render() {
    if (!this.state.loginSuccess) {
      return (
        <>
          <AzureLoginView
            azureInstance={this.azureInstance}
            loadingMessage={<Text style={{ fontSize: 20, fontWeight: "bold" }}>עוד רגע...מבצע אימות משתמש</Text>}
            onSuccess={this._onLoginSuccess}
          />
          <View style={styles.appBottomMainContainer}>
            <View style={styles.appBottomView}>
              <Text style={{ fontWeight: "bold" }}>Developed by Mobile Team (Mekorot) {'\u00A9'} 2020</Text>
            </View>
          </View>
        </>
      )
    }


    const { userPrincipalName, givenName } = this.state.azureLoginObject;
    console.log("zubak====", givenName)
    return (

      <PlacesNavigator />,
      <Modal />

    );

  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'center',
  },
  text: {
    textAlign: 'center',
    color: '#333333',
    marginBottom: 5
  },
  appBottomMainContainer:
  {

    alignItems: 'center',
    justifyContent: 'center',
    paddingTop: (Platform.OS === 'ios') ? 20 : 0,

  },
  appBottomView: {
    width: '100%',
    height: 30,
    backgroundColor: '#fc9208',
    justifyContent: 'center',
    alignItems: 'center',
    position: 'absolute',
    bottom: 0
  }
});
```

**Comment by tuzz** [comment]
I'd recommend asking on stackoverflow @roei133. This is the tracking issue for the specific problem where BackgroundFetch doesn't run after the app has been terminated or the phone has rebooted. It isn't a general support forum and most people here are busy with other things. Sorry.

**Comment by ioness** [comment]
> I'd recommend asking on stackoverflow @roei133. This is the tracking issue for the specific problem where BackgroundFetch doesn't run after the app has been terminated or the phone has rebooted. It isn't a general support forum and most people here are busy with other things. Sorry.

I have fear of being out of the topic but is this related to "silent" push notifications ?

Those:
1) Do not need user permissions to be sent.
2) Wake up the app and you have 30 seconds to fetch.

**Comment by lucas123ho** [comment]
For me it's working like this:

**./src/services/tasks.ts**
```javascript
import * as BackgroundFetch from 'expo-background-fetch';
import * as TaskManager from 'expo-task-manager';

export async function registerFetchTask(FETCH_TASKNAME: string, runBackgroundSaga, INTERVAL: number) {
    TaskManager.defineTask(FETCH_TASKNAME, runBackgroundSaga);

    const status = await BackgroundFetch.getStatusAsync();
    switch (status) {
        case BackgroundFetch.Status.Restricted:
        case BackgroundFetch.Status.Denied:
            console.log("Background execution is disabled");
            return;

        default: {
            console.debug("Background execution allowed");

            let tasks = await TaskManager.getRegisteredTasksAsync();
            if (tasks.find(f => f.taskName === FETCH_TASKNAME) == null) {
                console.log("Registering task");
                await BackgroundFetch.registerTaskAsync(FETCH_TASKNAME, {
                    minimumInterval: INTERVAL,
                    stopOnTerminate: false,
                    startOnBoot: true
                  },).then(() => BackgroundFetch.setMinimumIntervalAsync(INTERVAL));

                tasks = await TaskManager.getRegisteredTasksAsync();
                console.debug("Registered tasks", tasks);
            } else {
                console.log(`Task ${FETCH_TASKNAME} already registered, skipping`);
            }
        }
    }
}
```

**./App.tsx**
```javascript
import React, { useEffect, useCallback } from 'react';
import Home from './src/pages/Home';

import {registerFetchTask} from './src/services/tasks';
const INTERVAL_TASKS = 60000;

registerFetchTask('teste', () => {
  console.log('Rodando em background');
}, INTERVAL_TASKS);

export default function App() {
  const verifyChangesOnline = useCallback(() => {
    setInterval(() => {
      console.log('Verificando mudanca')
    }, INTERVAL_TASKS);
  }, []);

  useEffect(() => {
    verifyChangesOnline();
  }, [verifyChangesOnline]);


  return (
    <Home />
  );
}
```

**Comment by jeromecornet** [comment]
@lucas123ho thanks... when you say it works, in which scenarios from https://github.com/expo/expo/issues/3582#issuecomment-571695638 does it work ?

So far for me in the last released SDK (36) it is still broken for the last 2 cases. I've seen some code which may help for the next SDK release, is this what you are testing with ?

**Comment by lucas123ho** [comment]
@jeromecornet this is my scenarios:

**dependences:**
```json
"expo": "36.0.0",
"expo-background-fetch": "8.0.0",
"expo-task-manager": "8.0.0",
```
**SDK version:**
```json
"sdkVersion": "36.0.0",
```

I tested it on the Iphone 11 Pro Max simulator and my android device. Had different behaviors, in the simulator setInterval was executed in the background. On the device, what was being executed was the task.

It works for me. But the task dies when the app is removed from recent apps. I don't know if this is normal behavior.

**Comment by jeromecornet** [comment]
OK, makes sense @lucas123ho.... the fact that the task dies when the app is removed from recent apps is the problem, because it means you cannot rely on this to wake up the app when needed.

I'm not sure if @mczernek  or @tsapeta have had a chance to test out the minimal project to reproduce the scenario https://github.com/jeromecornet/expo-android-backgroundfetch  on the upcoming SDK after their latest changes but hopefully SDK37 will address this issue.

**Comment by calvinkei** [comment]
any idea when it will be fixed? from expo.canny.io it is not planned on the roadmap...

**Comment by mczernek** [comment]
> @jeromecornet
> It works for me. But the task dies when the app is removed from recent apps. I don't know if this is normal behavior.
This actually is normal and desired behavior. iOS assumes, that if you kill and app intentionally, you don't want it to perform any actions. Hence, app does not receive BackgroundFetch triggers. We do not have any control over that, and the behavior is identical with what pure native apps are facing.

I'am about to test your example on our recent code.

**Comment by mskg** [comment]
> > @jeromecornet
> > It works for me. But the task dies when the app is removed from recent apps. I don't know if this is normal behavior.
> 
> This actually is normal and desired behavior. iOS assumes, that if you kill and app intentionally, you don't want it to perform any actions. Hence, app does not receive BackgroundFetch triggers. We do not have any control over that, and the behavior is identical with what pure native apps are facing.

I can also confirm this.

**Comment by jeromecornet** [comment]
@mczernek thanks for pushing on the Android front, and @mskgthanks for the info. 

A couple of things on iOS: the new scheduler in iOS 13 https://developer.apple.com/documentation/backgroundtasks/bgtaskscheduler specifically mentions waking up the app at periodic intervals which may provide a better framework for this.

For iOS 12 and before, can you point me to the Apple documentation that describe this behavior ? This does not seem consistent with the fact that Xcode provides a special “Launch due to a background fetch event”, which should not be needed if BackgroundFetch tasks run only when the app is backgrounded and not terminated. I have not been able to find a proper reference in the documentation describing what happens after the phone reboots or after an app gets removed from the recent apps.

**Comment by mczernek** [comment]
I am not really sure this is documented at all, but this is knowledge
widely spread across community and confirmed with many people's experience,
including mine. This presentation describes more details on how Background
Fetch actually works: https://www.slideshare.net/moliver816/background-fetch

As of Xcode option, it makes perfect sense. Your app might be launched by
Background Fetch event, unless user killed it intentionally. Which means,
that if SYSTEM decides to kill your app (due to f.ex. running out of RAM)
then it will be launched again by Background Fetch. But when USER kills it,
it won't.

We are discussing switching to new BGTaskScheduler API from iOS13, since we
really like it. Moreover, similar API is present on Android for some time.
But this won't be available on SDK37, and the discussion about scope of
SDK38 is still in progress. Stay in touch with us and take a look at SDK38
milestone once it will be created.

czw., 20 lut 2020 o 15:24 Jerome Cornet <notifications@github.com>
napisał(a):

**Comment by roei133** [comment]
I wanna understand something please ...i need to use background service task that will keep working also when the app is kill , so why this service can work with regular cli (with no expo) and in expo it doesnt work ? 
i really dont understand it, and i really want that someone will explain me why it profitable to use with expo if i cant get background task work when the app is kill and who are use with regular cli so he can use task in background when the app is kill .

**Comment by pxmage** [comment]
This is really frustrating after I tried all the code above and checked expo docs for over 10 times at least still couldn't make it work on my device nor the simulator.

If expo team could provide a snack or an example of `TaskManager` or  `BackgroundFetch`, it will save a ton of time for me.

**Comment by mehmetsalihakcan** [comment]
Hi friends, i have same problem .i want to run when my app in backround or killed but it runs just app on foreground also setMinimumIntervalAsync is not working.Did anybody solve this problem?

**Comment by bobymicroby** [comment]
@mczernek  👍 Upvote for more reliable background tasks via BGTaskScheduler

**Comment by ugurdnlr** [comment]
I have the same problem. Have there been any developments?

**Comment by harveyjo** [comment]
+1 on needed this to be more reliable. Having the same issues many have described above after checking many sources of documentation and many forums. Have been developing an app for months that is basically done besides needing this feature to work, ASAP.

**Comment by mczernek** [comment]
> This is really frustrating after I tried all the code above and checked expo docs for over 10 times at least still couldn't make it work on my device nor the simulator.
> 
> If expo team could provide a snack or an example of `TaskManager` or `BackgroundFetch`, it will save a ton of time for me.

There is example for virtually every module we provide available within our repository. Here's one for [BackgroundFetch(https://github.com/expo/expo/blob/master/apps/native-component-list/src/screens/BackgroundFetchScreen.tsx).
There might seem obsolete sometimes (no hooks and so on) but we test it regularly so they should, most of the time, show each module works and should be used.


> +1 on needed this to be more reliable. Having the same issues many have described above after checking many sources of documentation and many forums. Have been developing an app for months that is basically done besides needing this feature to work, ASAP.

We are having it on our map. Exact timeline is difficult to predict, I am afraid. It won't, however, fix most pressing issue for you, which is executing tasks even after app is being swiped from recents. This is, and probably will be, system limitation.

As of now, sine BackgroundFetch seems to be working with system limitations, let me close this issue for now. For maintaining order, please report feature request if you wish to extend background capabilities for expo.

**Comment by juancardlm** [comment]
To make background tasks work, I used the **foregroundService** option for the **startLocationUpdatesAsync** method as it says in the docs:

_Use this option to put the location service into a foreground state, which will make location updates in the background as frequent as in the foreground state. As a downside, it requires a sticky notification, so the user will be aware that your app is running and consumes more resources even if backgrounded. **(Available since Android 8.0)**_

This is how it looks like on my code:

```
Location.startLocationUpdatesAsync(LOCATION_TASK_NAME, { 
    accuracy: Location.Accuracy.BestForNavigation,
    timeInterval: LOCATION_TIME_INTERVAL,
    distanceInterval: LOCATION_DISTANCE_INTERVAL,
    foregroundService: {
        notificationTitle: LOCATION_TITLE,
        notificationBody: LOCATION_SUBTITLE
    }
});

BackgroundFetch.registerTaskAsync(LOCATION_TASK_NAME, {
        minimumInterval: FETCH_INTERVAL,
        stopOnTerminate: false,
        startOnBoot: true
},).then(() => BackgroundFetch.setMinimumIntervalAsync(FETCH_INTERVAL));
```

I hope it helps!

**Comment by juancardlm** [comment]
> How are you retrieving the data from the location updates? In the docs it said to use TaskManager and define a task but that is not working for me.

This is how I define the task, just before the methods above.
Remember it cannot be called in any of React lifecycle methods like componentDidMount
```
TaskManager.defineTask(LOCATION_TASK_NAME, ({ data, error }) => {
     if (error)
        console.log("Error getting user location: ${JSON.stringify(error)}");
    else {
        const coords = data.locations[0].coords;
        console.log("User location retrieved: [${coords.latitude}, ${coords.longitude}]");
    }
});
```

If you have any problems getting current location, you can also use this code in the task function

```
navigator.geolocation.getCurrentPosition(
    (position) => {
        console.log(`Position retrieved: [${JSON.stringify(position)}]`);                               
    }, 
    (error) => console.log(`Unable to get position ${JSON.stringify(error)}`)
);
```

**Comment by Shpadoinkle** [comment]
Just want to share my experience with this.  I have spent two and half days trying to get this to work on iOS.  followed all the examples of code people had posted.  But can confirm nothing worked until i read @mskg comment that he HAD to fire Location.startLocationUpdatesAsync.  

Once i fired that from outside, everything started working, even on the simulator. 

Note that this is a mash of everyones code until something worked.. will clean up later 

I can't really shed more light on why it started working for me, and if i even wrote it correctly.. but all i can say is my background location tracking now works and hits my backend., but hopefully my code dump below will help out others


```

TaskManager.defineTask("locationTask", async ({ data, error }) => {
  try {
    if (error) {
      console.log(error)
      console.log("Background locationTask data error")
      return BackgroundFetch.Result.Failed
    }
    if (data) {
      const { locations } = data
      // do your logic thing here!!!
      return BackgroundFetch.Result.NewData
    }
  } catch (err) {
    console.log("Error in location in locationTask")
    return BackgroundFetch.Result.Failed
  }
})


 const registerFetchTask = async () => {
  const status = await BackgroundFetch.getStatusAsync()
  switch (status) {
    case BackgroundFetch.Status.Restricted:
    case BackgroundFetch.Status.Denied:
      console.log("Background execution is disabled")
      return

    default: {
      console.debug("Background execution allowed")
      let tasks = await TaskManager.getRegisteredTasksAsync()
      if (tasks.find(f => f.taskName === "locationTask") == null) {
        console.log("Registering task")
        await BackgroundFetch.registerTaskAsync("locationTask", {
          minimumInterval: 900,
          stopOnTerminate: false,
          startOnBoot: true,
        })

        tasks = await TaskManager.getRegisteredTasksAsync()
        console.debug("Registered tasks", tasks)
      } else {
        console.log(`Task ${"locationTask"} already registered, skipping`)
      }
      await BackgroundFetch.setMinimumIntervalAsync(900)
    }
  }
}

registerFetchTask()

Location.startLocationUpdatesAsync("locationTask", {
  accuracy: Location.Accuracy.Balanced,
  showsBackgroundLocationIndicator: true,
  timeInterval: 90000,
  activityType: Location.ActivityType.Other,
})

```

**Comment by obo20** [comment]
> > So, you have to
> > 
> > 1. Always call `TaskManager.defineTask`.
> > 2. Check if task is registered with `BackgroundFetch`. If not, register.
> > 3. Always call `BackgroundFetch.setMinimumIntervalAsync`.
> 
> re: 2. - there is no need to check if background fetch task is already registered - if you try to register it again, then it just overwrites the previously registered task, which is fine 👌 Except that, these steps look good.
> 
> > adding this at the end appears to make it work, as suggested by @tomas1000r why? @tsapeta
> > ```
> >     Location.startLocationUpdatesAsync(client.tasks.background)
> > ```
> 
> I'll try to check it out soon, but it seems weird to me as there is no connection between location and background fetch. They are completely independent features. My assumption is that, if you turn on background location then your app is actually always active, so the background fetch can occur in the time you set by `setMinimumIntervalAsync`. However, if the app is not active, then it strongly depends on the operating system and how you use the application. Apple has its own algorithm which defines how often the background fetch should trigger, based on your own usage of the app. If you use it a lot, then it will fetch as often as possible, but if you use it like at 4pm every day, the background fetch should trigger just before, so your data is updated when you launch it. So imho you shouldn't worry about the task not being done on time - Apple knows better when to call it 😅

@tsapeta Can you think of anything that can be done from within Expo that can replicate this 'keep active' functionality without requiring location updates constantly being run in the background? 

Unfortunately with iOS now actively showing at all times when an app is using locational data, having our app constantly checking the location of our users when it has no need to is a really bad look for our application.

**Comment by srikanthkh** [comment]
I can not get background fetch to work on iOS(when app is in background or not does not matter). I tried using the code from the comments but nothing seem to work. I was able to get the android version working with the same code though. Tried using expo 39 and 40 with all latest packages.

1. TaskManager.defineTask.
2. Check if task is registered with BackgroundFetch. If not, register.
3. call BackgroundFetch.setMinimumIntervalAsync.
Adding Location.startLocationUpdatesAsync(BACKGROUND_TASK) does seem to invoke the background task once.

Was anyone successful in getting background fetch to work without Location in iOS?? Please share code. Thanks for your help in advance.

**Comment by juancardlm** [comment]
> I can not get background fetch to work on iOS(when app is in background or not does not matter). I tried using the code from the comments but nothing seem to work. I was able to get the android version working with the same code though. Tried using expo 39 and 40 with all latest packages.
> 
> 1. TaskManager.defineTask.
> 2. Check if task is registered with BackgroundFetch. If not, register.
> 3. call BackgroundFetch.setMinimumIntervalAsync.
>    Adding Location.startLocationUpdatesAsync(BACKGROUND_TASK) does seem to invoke the background task once.
> 
> Was anyone successful in getting background fetch to work without Location in iOS?? Please share code. Thanks for your help in advance.

Have you tried `foregroundService` option for `Location.startLocationUpdatesAsync` method?
This is how it looks like on my code (works both Android and IOS): 

```
Location.startLocationUpdatesAsync(LOCATION_TASK_NAME, {
    accuracy: Location.Accuracy.BestForNavigation,
    timeInterval: LOCATION_TIME_INTERVAL,
    distanceInterval: LOCATION_DISTANCE_INTERVAL,
    foregroundService: {
        notificationTitle: LOCATION_TITLE,
        notificationBody: LOCATION_SUBTITLE
    }
});

**Comment by srikanthkh** [comment]
@juancardlm This does work.. thank you for your reply! But I am trying to get background fetch work without using expo Location. My app does not need Location so do not want to ask permission for that from the user.

**Comment by artsnr1** [comment]
> > This is really frustrating after I tried all the code above and checked expo docs for over 10 times at least still couldn't make it work on my device nor the simulator.
> > If expo team could provide a snack or an example of `TaskManager` or `BackgroundFetch`, it will save a ton of time for me.
> 
> There is example for virtually every module we provide available within our repository. Here's one for [BackgroundFetch](https://github.com/expo/expo/blob/master/apps/native-component-list/src/screens/BackgroundFetchScreen.tsx).
> There might seem obsolete sometimes (no hooks and so on) but we test it regularly so they should, most of the time, show each module works and should be used.
> 
> > +1 on needed this to be more reliable. Having the same issues many have described above after checking many sources of documentation and many forums. Have been developing an app for months that is basically done besides needing this feature to work, ASAP.
> 
> We are having it on our map. Exact timeline is difficult to predict, I am afraid. It won't, however, fix most pressing issue for you, which is executing tasks even after app is being swiped from recents. This is, and probably will be, system limitation.
> 
> As of now, sine BackgroundFetch seems to be working with system limitations, let me close this issue for now. For maintaining order, please report feature request if you wish to extend background capabilities for expo.

When would this be resolved. It is such a 'must-have' feature. I need to fetch data periodically and most of the time an app is not in the background/foreground

**Comment by srikanthkh** [comment]
@artsnr1 Does background fetch work for you when the app is in background/foreground? If Yes, do you use location?

**Comment by artsnr1** [comment]
@srikanthkh I haven't tried it yet as it doesn't provide for my need to work even on app termination

**Comment by rosnaib11** [comment]
If your app requires background location, the best solution is to eject so you can use native modules. [@mauron85/react-native-background-geolocation](https://github.com/mauron85/react-native-background-geolocation) or headless JS

**Comment by artsnr1** [comment]
Unfortunately, ejection is the last resort

**Comment by pxmage** [comment]
> This is really frustrating after I tried all the code above and checked expo docs for over 10 times at least still couldn't make it work on my device nor the simulator.
> 
> If expo team could provide a snack or an example of `TaskManager` or `BackgroundFetch`, it will save a ton of time for me.

Well, finally I ejected, any recommendations for the alternatives?

**Comment by rosnaib11** [comment]
To me, I used mauron combined with headlessJS. Because mauron location listener doesn't work if app is terminated. That's why I used headlessJS. But sending of location using postTemplate will work even if app is terminated. I used headlessJS just for my own purpose.

**Comment by trajano** [comment]
I have an MVCE that shows this problem https://github.com/trajano/expo-mvce it's based off https://github.com/expo/expo/blob/master/apps/native-component-list/src/screens/BackgroundFetchScreen.tsx in my case it works on Android, but not iOS Expo clients.  I am using a custom Expo client as well

I also added BackgroundFetch.setMinimumIntervalAsync(60);

**Comment by sicero** [comment]
Why was this issue closed if no resolution - it seems such a fundamental feature of BackGround fetch? What is the alternative?

**Comment by LeifS** [comment]
Was a solution ever found for this problem? 
Or is the only way to get BackgroundFetch to work on iOS to use the hacky Location.startLocationUpdatesAsync() work around? The problem with this fix is that it runs the task in the foreground as well as the background, which causes new issues.

**Comment by trajano** [comment]
@LeifS  @sicero  it does work, my MVCE just needs to update the README but the MVCE does work.  There's one caveat, on iOS you can't really predict when the background fetch will fire.  You need to keep the phone fully charged and plugged into get it more frequent https://github.com/trajano/expo-mvce

**Comment by marksalpeter** [comment]
BackgroundFetch doesn’t work at all for me on the simulator, in expo go, or in the compiled app. The task is defined in the list of tasks, it’s properly started, and the background fetch never fires - not one time. It fails completely silently with no explanation as to why. 

I’ve abandoned expo as a framework because of the lack of support for this feature and because of how poorly they’ve handled this matter. I strongly recommend that other do the same. This framework is extremely unreliable and the support for it is nonexistent.

**Comment by rikur** [comment]
@marksalpeter I just did the opposite: I'm ditching semi-abandoned volunteer-run react-native libraries and been enjoying Expo unimodules a lot more. Best of luck!

**Comment by brentvatne** [comment]
for anyone that stumbles on this - @ajsmth recently wrote up some useful instructions for trigger background fetch manually, which is much easier than waiting for the operating system to kick in for you: https://docs.expo.io/versions/latest/sdk/background-fetch/#triggering-background-fetches

i'm going to lock this thread because as far as we know this is working as expected, please make a new issue if you continue to encounter issues.
